cpu/cortexm_common/thread_arch.c

```diff
@@ -296,84 +308,117 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
+
+    "pop    {lr}                      \n" /* Pop exception from the exception stack */
+
+    "mov    r1,r12                    \n" /* r1 = sched_active_thread */
+    "cbz    r1, restore_context       \n" /* goto restore_context if r1 == 0 */
+
+    "mrs    r2, psp                   \n" /* get stack pointer from user mode */
+
+#ifdef MODULE_CORTEXM_FPU
+    "tst    lr, #0x10                 \n"
+    "it     eq                        \n"
+    "vstmdbeq r2!, {s16-s31}          \n" /* save FPU registers if FPU is used */
+#endif
+    "stmdb  r2!,{r4-r11,lr}           \n" /* save regs, including lr */
+    "str    r2, [r1]                  \n" /* write r0 to thread->sp */
+
+    /* current thread context is now saved */
+    "restore_context:                 \n" /* Label to skip thread state saving */
+
+    "ldr    r0, [r0]                  \n" /* load tcb->sp to register 1 */
+    "ldmia  r0!, {r4-r11,lr}          \n" /* restore other registers, including lr */
+#ifdef MODULE_CORTEXM_FPU
+    "tst    lr, #0x10                 \n"
+    "it     eq                        \n"
+    "vldmiaeq r0!, {s16-s31}          \n" /* load FPU registers if saved */
+#endif
+    "msr    psp, r0                   \n" /* restore user mode SP to PSP reg */
+    "bx     lr                        \n" /* load exception return value to PC,
+                                           * causes end of exception*/
+#else /* CPU_CORE_CORTEXM_FULL_THUMB */
+
+    /* Cortex-M0(+) and Cortex-M23 */
+    "cmp    r0, r12                   \n" /* if r0 == previous_thread: */
+    "bne    cont_schedule             \n" /*   jump over pop if r0 != 0 */
+    "pop    {pc}                      \n" /*   Pop exception return to PC */
+
+    "cont_schedule:                   \n" /* Otherwise continue the ctx switch */
+
+    "pop    {r2}                      \n" /* Pop LR from the exception stack */
+    "mov    lr, r2                    \n" /* Store LR in lr */
+    "mov    r1,r12                    \n" /* r1 = sched_active_thread */
+    "cmp    r1, #0                    \n" /* Test if r1 == NULL */
+    "mov    r12, sp                   \n" /* remember the exception SP in r12 */
+    "beq    restore_context           \n" /* goto restore_context if r1 == NULL */
 
     "mrs    r0, psp                   \n" /* get stack pointer from user mode */
-#if defined(CPU_CORE_CORTEX_M0) || defined(CPU_CORE_CORTEX_M0PLUS) \
-    || defined(CPU_CORE_CORTEX_M23)
-    "push   {r1}                      \n" /* push sched_active_thread */
-    "mov    r12, sp                   \n" /* remember the exception SP */
     "mov    sp, r0                    \n" /* set user mode SP as active SP */
+
+    /* Calculate the expected stack offset beforehand so that we don't have to
+     * store the old SP from here on, saves a register we don't have */
+#if defined (__clang__)
+    /* Clang doesn't like the 'sub' instruction, swap for a subs. Generates
```

Comment:
try this: https://gist.github.com/1344638dd2ad82b5c2019f71dea639cb

Reply:
Awesome, thanks!